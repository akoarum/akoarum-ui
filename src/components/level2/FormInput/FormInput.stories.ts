import { storiesOf } from '@storybook/vue'
import { action } from '@storybook/addon-actions'
import { withKnobs, text, select, boolean } from '@storybook/addon-knobs'
import { withInfo } from 'storybook-addon-vue-info'
import FormInput from './FormInput.vue'

storiesOf('level2/FormInput', module)
  .addDecorator(withInfo)
  .addDecorator(withKnobs)
  .add('Normal', () => ({
    components: { FormInput },
    props: {
      type: { default: select('type属性値', {
        text: 'text',
        number: 'number',
        email: 'email',
        tel: 'tel',
        password: 'password'
      }, 'text') },
      min: { default: text('min', '') },
      max: { default: text('max', '') },
      minlength: { default: text('minlength', '') },
      maxlength: { default: text('maxlength', '') },
      required: { default: boolean('必須項目', true) },
      placeholder: { default: text('プレスホルダー', '') },
      width: { default: text('width', '') },
    },
    data() {
      return {
        value: ''
      }
    },
    template: `
      <FormInput
        v-model="value"
        :type="type"
        :min="min ? Number(min) : null"
        :max="max ? Number(max) : null"
        :minlength="minlength ? Number(minlength) : null"
        :maxlength="maxlength ? Number(maxlength) : null"
        :placeholder="placeholder"
        :required="required"
        :width="width"
        name="name"
        @focus="focus"
        @blur="blur"
        @error="error"
        @dirty="dirty"
      />
    `,
    methods: {
      focus: action('focus'),
      blur: action('blur'),
      error: action('error'),
      dirty: action('dirty')
    }
  }), { info: {} })
