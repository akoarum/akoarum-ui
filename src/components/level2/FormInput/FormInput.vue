<template>
  <div class="formInput">
    <AInput
      v-model="valueModel"
      v-bind="$attrs"
      :id="name"
      :type="type"
      :name="name"
      :required="required"
      :min="min"
      :max="max"
      :minlength="minlength"
      :maxlength="maxlength"
      :aria-invalid="String(!!error)"
      :aria-describedby="error ? `${name}-error` : false"
      :width="width"
      class="formInput__input"
      @focus="handleFocus"
      @blur="handleBlur"
    />
    <template v-if="error">
      <ATexts
        :id="`${name}-error`"
        :aria-live="error ? 'assertive' : 'off'"
        role="alert"
        size="14"
        class="formInput__error"
        caution
      >
        <span>{{ error }}</span>
      </ATexts>
    </template>
  </div>
</template>

<script lang="ts">
import {
  defineComponent,
  watch,
  onMounted,
  SetupContext
} from '@vue/composition-api'
import { useFormValidation } from '~/compositions/useFormValidation'
import { ATexts, AInput } from '~/components/level1'

interface Props {
  name: string
  value: string
  type: string
  width?: string
  required?: boolean
  min?: number
  max?: number
  minlength?: number
  maxlength?: number
}

interface State {
  error: string
  isDirty: boolean
}

export default defineComponent({
  inheritAttrs: false,
  components: { ATexts, AInput },
  props: {
    name: { type: String, required: true },
    value: { type: String, required: true },
    type: { type: String, required: true },
    required: { type: Boolean },
    min: { type: Number },
    max: { type: Number },
    minlength: { type: Number },
    maxlength: { type: Number },
    width: { type: String }
  },
  setup(props: Props, context: SetupContext) {
    const {
      error,
      isDirty,
      validation,
      valueModel,
      handleFocus,
      handleBlur
    } = useFormValidation(props, context)

    onMounted(() => {
      if (props.value) {
        error.value = validation(props.value)
      }
    })

    watch(error, () => {
      context.emit('error', !!error)
    })

    watch(isDirty, () => {
      context.emit('dirty', true)
      error.value = validation(props.value)
    })

    return { error, isDirty, validation, valueModel, handleFocus, handleBlur }
  }
})
</script>

<style lang="scss" scoped>
.formInput__error {
  margin-top: 5px;
}
</style>
