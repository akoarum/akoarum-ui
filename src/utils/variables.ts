export const FORM_VALIDATE_ERRORS = {
  REQUIRED: 'この項目は入力必須です。',
  EMAIL: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスの形式になっていません。'
  },
  TEL: {
    regex: /^\d{2,5}-?\d{1,4}-?\d{4}$/,
    message: '電話番号（半角数字）で入力してください。'
  }
}
