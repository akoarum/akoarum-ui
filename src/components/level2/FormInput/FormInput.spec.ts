import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueCompositionApi from '@vue/composition-api'
import FormInput from './FormInput.vue'
import { FORM_VALIDATE_ERRORS } from '~/utils/variables'

describe('FormInput', () => {
  const localVue = createLocalVue()
  let wrapper

  localVue.use(VueCompositionApi)

  beforeEach(() => {
    wrapper = shallowMount(FormInput, {
      propsData: {
        name: 'name',
        value: '',
        type: 'text'
      },
      localVue
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })
    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('event', () => {
    it('フォーカス時に親に通知する', () => {
      wrapper.vm.handleFocus()
      expect(wrapper.emitted().focus).toBeTruthy()
    })

    it('フォーカスが外れたことを通知する', () => {
      wrapper.vm.handleBlur()
      expect(wrapper.emitted().blur).toBeTruthy()
    })
  })

  describe('validation', () => {
    it('requiredで未入力の場合は未入力エラーになる', () => {
      wrapper.setProps({ required: true })
      wrapper.setData({ isDirty: true })
      expect(wrapper.vm.validation('')).toBe(FORM_VALIDATE_ERRORS.REQUIRED)
    })

    it('type: emailでメールアドレスの形式になっていなければエラーになる', () => {
      wrapper.setProps({ type: 'email' })
      wrapper.setData({ isDirty: true })
      expect(wrapper.vm.validation('aaa')).toBe(FORM_VALIDATE_ERRORS.EMAIL.message)
    })

    it('type: telで電話番号の形式になっていなければエラーになる', () => {
      wrapper.setProps({ type: 'tel' })
      wrapper.setData({ isDirty: true })
      expect(wrapper.vm.validation('111')).toBe(FORM_VALIDATE_ERRORS.TEL.message)
    })

    it('type: numberでminより下になっていたらエラーになる', () => {
      wrapper.setProps({ type: 'number', min: 10 })
      wrapper.setData({ isDirty: true })
      expect(wrapper.vm.validation('1')).toBe('10以上にしてください。')
    })

    it('type: numberでmaxより上になっていたらエラーになる', () => {
      wrapper.setProps({ type: 'number', max: 10 })
      wrapper.setData({ isDirty: true })
      expect(wrapper.vm.validation('11')).toBe('10以下にしてください。')
    })

    it('文字数がminlengthより少なかったらエラーになる', () => {
      wrapper.setProps({ minlength: 10 })
      wrapper.setData({ isDirty: true })
      expect(wrapper.vm.validation('a')).toBe('10文字以上入力してください。')
    })

    it('文字数がmaxlengthより多かったらエラーになる', () => {
      wrapper.setProps({ maxlength: 4 })
      wrapper.setData({ isDirty: true })
      expect(wrapper.vm.validation('aaaaa')).toBe('4文字以下にしてください。')
    })
  })
})
