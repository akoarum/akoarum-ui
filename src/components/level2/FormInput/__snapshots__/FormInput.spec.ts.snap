// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput initialized snapshot 1`] = `
<div
  class="formInput"
>
  <ainput-stub
    aria-invalid="false"
    class="formInput__input"
    id="name"
    name="name"
    type="text"
    value=""
  />
   
  <!---->
</div>
`;
